[naga msl-out] Emit and init `struct` member padding always

ErichDonGubler opened the pull request:
**Checklist**

- [x] ~~Run `cargo clippy`.~~
- [x] ~~Run `cargo clippy --target wasm32-unknown-unknown` if applicable.~~
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
_Link to the issues addressed by this PR, or dependent PRs in other repositories_

Resolves #4522.

**Description**
_Describe what problem this is solving, and how it's solved._

Previously, implicit padding members of `struct`s were suppressed from structure definitions in Metal output if they had a binding specified for them (i.e., `@location(0)`). This padding is, however, is necessary for correct access of member fields passed into shaders by uniform and storage buffers. Unconditionally emit padding members for `struct`s.

**Testing**
_Explain how this change is tested._

I've added a new `struct-layout` test to `naga`'s `snapshots` test suite to exercise the adjustments to vertex shader coercion of `varyings` structures (i.e., "

ErichDonGubler committed:
fix(msl-out): emit and init `struct` member padding always

Previously, implicit padding members of `struct`s were suppressed from
structure definitions in Metal output if they had a binding specified
for them (i.e., `@location(0)`). This padding is, however, is necessary
for correct access of member fields passed into shaders by uniform and
storage buffers. Unconditionally emit padding members for `struct`s.

Resolves
[`gfx-rs/wgpu`#4701](https://github.com/gfx-rs/wgpu/pull/4701).

ErichDonGubler committed:
chore(msl-out): `should_pack_struct_member`: remove unnecessary alignment check for `vec3`s

ErichDonGubler committed:
chore(msl-out): remove outdated "quick and dirty" comment

teoxoy commented:
@ErichDonGubler and I went over this but ultimately found out that this solution might cause more issues than it solves.

I filed https://github.com/gfx-rs/wgpu/issues/4703 to track the full solution.

ErichDonGubler commented:
Reopening this, as this fix is _actually_ the right direction for #4522 (which has been reopened, also).

ErichDonGubler committed:
test(naga): add `struct-layout` test

ErichDonGubler commented:
@teoxoy (who I presume will review this): I've added e70e4ba28, 5a8e4ae13, and a199623c7 to address what I believe is the remaining scope of work here. I'm uncertain if the test I've added needs more cases, so I've marked it as WIP for now. I can reword that commit if you think it's good enough, tho.

teoxoy commented on naga/CHANGELOG.md:
```diff
@@ -79,6 +79,7 @@ For changelogs after v0.14, see [the wgpu changelog](../CHANGELOG.md).
 
 - Add and fix minimum Metal version checks for optional functionality. ([#2486](https://github.com/gfx-rs/naga/pull/2486)) **@teoxoy**
 - Make varyings' struct members unique. ([#2521](https://github.com/gfx-rs/naga/pull/2521)) **@evahop**
+- Emit and init `struct` member padding always ([#4701](https://github.com/gfx-rs/wgpu/pull/4701)) **@ErichDonGubler**
```
This should go in the changelog at the root of the project.

ErichDonGubler replied:
oderp, fixed with a6c176f4f. 😅

teoxoy commented on naga/tests/out/msl/quad.msl:
```diff
@@ -6,6 +6,7 @@ using metal::uint;
 
 struct VertexOutput {
     metal::float2 uv;
+    char _pad1[8];
```
Hmm, we don't need to explicitly pad here but I guess this doesn't hurt anything and is actually a preexisting issue.

The padding calculation needs to align the offset before it checks the member's offset.
We can fix it later.

teoxoy reviewed (approved):
Great stuff!
